lib/remote_node.go

```diff
@@ -0,0 +1,468 @@
+func (rn *RemoteNode) HandleVersionMessage(verMsg *MsgDeSoVersion, responseNonce uint64) {
+	if verMsg.Version < rn.params.MinProtocolVersion {
+		glog.V(1).Infof("RemoteNode.HandleVersionMessage: Requesting PeerDisconnect for id: (%v) "+
+			"protocol version too low. Peer version: %v, min version: %v", rn.peer.ID, verMsg.Version, rn.params.MinProtocolVersion)
+		rn.Disconnect()
+		return
+	}
+
+	vMeta := rn.GetHandshakeMetadata()
+	if vMeta.versionTimeExpected != nil && vMeta.versionTimeExpected.Before(time.Now()) {
+		glog.V(1).Infof("RemoteNode.HandleVersionMessage: Requesting PeerDisconnect for id: (%v) "+
+			"version timeout. Time expected: %v, now: %v", rn.peer.ID, vMeta.versionTimeExpected.UnixMicro(), time.Now().UnixMicro())
+		rn.Disconnect()
+		return
+	}
+
+	// Save the version nonce so we can include it in our verack message.
+	vMeta.versionNoncesReceived = verMsg.Nonce
+
+	// Set the peer info-related fields.
+	vMeta.userAgent = verMsg.UserAgent
+	vMeta.ServiceFlag = verMsg.Services
+	vMeta.advertisedProtocolVersion = NewProtocolVersionType(verMsg.Version)
+	negotiatedVersion := rn.params.ProtocolVersion
+	if verMsg.Version < rn.params.ProtocolVersion.ToUint64() {
+		negotiatedVersion = NewProtocolVersionType(verMsg.Version)
+	}
+	vMeta.negotiatedProtocolVersion = negotiatedVersion
+	vMeta.minTxFeeRateNanosPerKB = verMsg.MinFeeRateNanosPerKB
+	timeConnected := time.Unix(verMsg.TstampSecs, 0)
+	vMeta.timeConnected = &timeConnected
+	vMeta.timeOffsetSecs = verMsg.TstampSecs - time.Now().Unix()
+	vMeta.StartingBlockHeight = verMsg.StartBlockHeight
+
+	// Update the timeSource now that we've gotten a version message from the peer.
+	rn.cmgr.AddTimeSample(rn.peer.Address(), timeConnected)
+
+	if !rn.peer.IsOutbound() {
+		// Respond to the version message if this is an inbound peer.
+		rn.sendVersionMessage(responseNonce)
+	}
+	// After sending and receiving a compatible version, complete the
+	// negotiation by sending and receiving a verack message.
+	verackTimeExpected := time.Now().Add(rn.params.VersionNegotiationTimeout)
+	vMeta.verackTimeExpected = &verackTimeExpected
+	if err := rn.sendVerack(); err != nil {
+		glog.Errorf("RemoteNode.HandleVersionMessage: Problem sending verack message to peer (id= %d): %v", rn.peer.ID, err)
+		rn.Disconnect()
+		return
+	}
+}
+
+func (rn *RemoteNode) sendVerack() error {
+	verackMsg, err := rn.newVerackMessage()
+	if err != nil {
+		return err
+	}
+
+	if err := rn.cmgr.SendMessage(verackMsg, rn.peer.ID); err != nil {
+		glog.Errorf("RemoteNode.SendVerack: Problem sending verack message to peer (id= %d): %v", rn.peer.ID, err)
+		return err
+	}
+	return nil
+}
+
+func (rn *RemoteNode) newVerackMessage() (*MsgDeSoVerack, error) {
+	verack := NewMessage(MsgTypeVerack).(*MsgDeSoVerack)
+	vMeta := rn.GetHandshakeMetadata()
+
+	// Include the nonce we received in the peer's version message
+	verack.NonceReceived = vMeta.versionNoncesReceived
+	if vMeta.negotiatedProtocolVersion == ProtocolVersion2 {
```

Comment:
not a issue for this PR but do we want `negotiatedProtocolVersion` to actually represent the release tag of core that node is running? This may be useful for future forks.

Reply:
Yeah, I think this is a good idea!